keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+  as a lightweight OIDC proxy to connect many providers to Kubernetes
+- Change authenticator settings without restarting kube-apiserver
+
+### Non-Goals
+
+- Monitoring
+
+## Proposal
+
+1. Add new `authentication` API object to parse a structured config file `OIDCConfiguration`.
+2. Add a single flag for kube-apiserver to point to the structured config file, automatically reload it on changes.
+3. Use an expression language to let users write their own logic for mappings and validation rules 
+  (expressions should be simple for common cases, yet powerful to cover most user stories).
+
+### Risks and Mitigations
+
+Since this is a new optional feature, no migration is required. Before the Stable release of the feature,
+we should provide examples of migrating from a flag-based config to a new structured config.
+
+## Design Details
+
+### Configuration file
+
+The main part of this proposal is a configuration file. It contains an array of providers:
+
+```go
+type OIDCConfiguration struct {
+    metav1.TypeMeta
+    // Providers is a list of OIDC providers to authenticate Kubernetes users.
+    Providers []Provider
+}
+```
+
+Each provider has several properties that will be described in detail below.
+
+```go
+type Provider struct {
+    // Issuer is a basic OIDC provider connection options.
+    Issuer Issuer
+    // ClaimValidationRules are rules that are applied to validate ID token claims to authorize users.
+    // +optional
+    ClaimValidationRules []ClaimValidationRule
+    // ClaimMappings points claims of an ID token to be treated as user attributes.
+    // All mappings are logical expressions that is written in CEL https://github.com/google/cel-go.
+    ClaimMappings UserAttributes
+    // ClaimsFilter allows unmarshalling only required claims which positively affects performance.
+    // +optional
+    ClaimsFilter []string
+}
+```
+
+1. `Issuer` - is a sections for external provider specific settings, e.g., OIDC discovery URL. 
+  There is also default validation settings (ClientID, SkipOIDCValidations) that's available out of the box for flag-based 
+  OIDC provider, but for structured config they can be disabled.
+
+    ```go
+    type Issuer struct {
+        // URL points to the issuer URL in a format schema://url/path.
+        URL string
+        // CertificateAuthorityData contains PEM-encoded certificate authority certificates. Overrides CertificateAuthority
+        CertificateAuthorityData []byte
+        // ClientID the JWT must be issued for, the "sub" field. This plugin only trusts a single
+        // client to ensure the plugin can be used with public providers.
+        // Do not affect anything with the SkipOIDCValidations option enabled.
+        // +optional
+        ClientID string
+        // SkipOIDCValidations is a flag to turn off issuer validation, client id validation.
+        // OIDC related checks.
+        // +optional
+        SkipOIDCValidations bool
```

Comment:
why would we allow skipping these validations? this doesn't seem like something we should do

Reply:
Reminder: this might be better as ~3 separate API fields that define the default for each specific validation.  When unset, you would get the regular OIDC ID token semantics.

Reply:
talked about this offline... if we want this authenticator/config to support JWTs that aren't OIDC-compliant, we should rename the flag/config type (and maybe even the KEP) to be something other than OIDC

Reply:
It can be named `JWT authenticator`, but I'm still unsure because it only works with the OIDC discovery endpoint (issuer URL is required, and it tries to get the endpoint with public keys using the OIDC discovery mechanism).

Should we also give users a way to provide a direct endpoint to fetch keys, e.g., 
`keysEndpoint: https://my-oidc/keys` or `keysEndpoint: file://keys.json`?

Reply:
> but I'm still unsure because it only works with the OIDC discovery endpoint

IMO this is a non-issue.  Kubernetes SA accounts use OIDC discovery to distribute keys, but otherwise do the bare minimum to have ID token like semantics.  But yes, we should rename the KEP to make it clear that all you need the JWT format and an OIDC type discovery endpoint for key distribution.